warehouse/integrations/snowflake/snowflake.go

```diff
@@ -521,63 +509,133 @@ func (sf *Snowflake) joinColumnsWithFormatting(columns []string, format string)
 	}, ",")
 }
 
-func (sf *Snowflake) sampleDuplicateMessages(
-	ctx context.Context,
-	db *sqlmw.DB,
-	mainTableName,
-	stagingTableName string,
-) ([]duplicateMessage, error) {
+func (sf *Snowflake) trackDuplicateMessageInfo(
+	ctx context.Context, db *sqlmw.DB, mainTableName,
+	stagingTableName, sortedColumns string,
+) error {
 	if !lo.Contains(sf.config.debugDuplicateWorkspaceIDs, sf.Warehouse.WorkspaceID) {
-		return nil, nil
+		return nil
 	}
 	if !lo.Contains(sf.config.debugDuplicateTables, mainTableName) {
-		return nil, nil
+		return nil
 	}
 
 	identifier := sf.schemaIdentifier()
 	mainTable := fmt.Sprintf("%s.%q", identifier, mainTableName)
 	stagingTable := fmt.Sprintf("%s.%q", identifier, stagingTableName)
 
-	rows, err := db.QueryContext(ctx,
-		`SELECT ID, RECEIVED_AT
-		FROM `+mainTable+`
-		WHERE
-		  	RECEIVED_AT > (SELECT DATEADD(day,-?,current_date())) AND
-		  	ID IN (
-				SELECT ID
-				FROM `+stagingTable+`
-		  	)
-        ORDER BY RECEIVED_AT ASC
-		LIMIT ?;`,
+	rows, err := db.QueryContext(ctx, `
+		WITH mainCTE AS (
+		  SELECT
+			`+sortedColumns+`
+		  FROM
+			`+mainTable+`
+		  WHERE
+			RECEIVED_AT > (
+			  SELECT
+				DATEADD(
+				  day,
+				  -?,
+				  current_date()
+				)
+			)
+			AND ID IN (
+			  SELECT
+				ID
+			  FROM
+				`+stagingTable+`
+			)
+		  ORDER BY
+			RECEIVED_AT ASC
+		  LIMIT
+			?
+		), StagingCTE AS (
```

Comment:
[optional] Do we really need `StagingCTE`?

* In `mainCTE` we're getting all the records from the last X days that are also present in the staging table.
* In `stagingCTE` we're reading from the staging table instead but we select only the records that are already set as duplicate in `mainCTE`.

I understand that in the `UNION` at the end you want the identifier table but what for? If those records are duplicated they are in both tables, right?

Just saying, maybe just save the `mainCTE` and then add the staging file in the dump unless you find this way to be more convenient.

Reply:
They will have the same `messageID`, not other columns, that's what we will be figuring out from the sample.

Reply:
Yeah that's why I said we could include the staging file in the dump, it's data we have already. I get it though, looks more convenient this way and the additional scanning is as big as the data from the staging file so it shouldn't be huge :+1: